event_sourcery_esdb/outbox.py

```diff
@@ -1,202 +1,85 @@
-    _filterer: OutboxFiltererStrategy
+    @property
+    def active_subscription(self) -> PersistentSubscription:
+        assert self._active_subscription is not None
+        return self._active_subscription
 
     def put_into_outbox(self, events: list[RawEvent]) -> None:
         ...
 
-    def outbox_entries(
-        self, limit: int
-    ) -> Iterator[Tuple[EntryId, RawEvent, StreamId]]:
-        with self.outbox() as outbox:
-            events = self._client.read_all(outbox.read_from_position)
-            return (
-                (
-                    cast(int, entry.commit_position),
-                    raw_event,
-                    stream.Name(stream_name=entry.stream_name).uuid,
-                )
-                for entry in events
-                if outbox.should_emit(entry.commit_position)
-                and self._filterer(raw_event := dto.raw_event(entry))
-            )
-
-    def decrease_tries_left(self, entry_id: EntryId) -> None:
-        self._client.append_event(
-            Outbox.name,
-            StreamState.ANY,
-            Outbox.FailedAttempt(position=Position(entry_id)),
-        )
+    def outbox_entries(self, limit: int) -> Iterator[ContextManager[RawEvent]]:
+        info = self._client.get_subscription_info(self.OUTBOX_NAME)
+        if info.live_buffer_count == 0:
+            return
 
-    def remove_from_outbox(self, entry_id: EntryId) -> None:
-        self._client.append_event(
-            Outbox.name,
-            StreamState.ANY,
-            Outbox.Emitted(position=Position(entry_id)),
-        )
+        with self._context(limit) as subscription:
+            try:
+                for entry in subscription:
+                    if self._filterer(event := dto.raw_event(entry)):
+                        yield self._publish_context(entry, event)
+            except DeadlineExceeded:
+                pass
 
     @contextmanager
-    def outbox(self) -> Iterator[Outbox]:
-        outbox = Outbox()
+    def _publish_context(
+        self,
+        entry: RecordedEvent,
+        event: RawEvent,
+    ) -> Generator[RawEvent, None, None]:
         try:
-            for event in self._client.read_stream(stream_name=outbox.name):
-                outbox.__apply__(event)
-        except NotFound:
-            pass
-        yield outbox
-        if outbox.changes > self.MAX_EVENTS_BEFORE_SNAPSHOT:
-            self._create_snapshot(outbox)
-
-    def _create_snapshot(self, outbox: Outbox) -> None:
-        self._client.append_event(
-            outbox.name,
-            StreamState.ANY,
-            outbox.make_snapshot(),
-        )
-        self._client.append_event(
-            outbox.metadata,
-            StreamState.ANY,
-            TruncateBeforeEvent(position=Position(outbox.version + 1)),
-        )
+            yield event
+        except Exception:
+            logger.exception("Failed to publish message #%d", entry.id)
+            if entry.retry_count and entry.retry_count >= 2:
```

Comment:
I assume this is an implementation of up to two retries? Is it possible to retrieve these events later?
In the case of SQL-based event store, they are left behind in the outbox table with retries = 0.

Reply:
Yes, it's up to 3 times, and we have tests for this.
Action that I choose is 
`                self.active_subscription.nack(entry.id, action="park")`
And Park events are available in ESDB to check and to retry. You can do this from UI of ESDB